### PR TITLE
OPENEUROPA-2025: Use custom image effect to prevent retina image reducction.

### PR DESCRIPTION
**Pull request by imanoleguskiza**
### Description

When the images we use are too small, the same image is used for both regular and retina displays. This means that on retina displays the image gets displayed at half size. 
Fix the issue by upscaling the image if it is smaller than the desired  scale.

### Change log

- Added: Custom retina scale effect
- Changed:
- Deprecated:
- Removed:
- Fixed:
- Security:

### Commands

```sh
[Insert commands here]

```

### CODE REVIEW

File: modules/oe_theme_helper/config/schema/oe_theme_helper.schema.yml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+image.effect.retina_image_scale:
+  type: image_size
```
Comment (upchuk) [inline modules/oe_theme_helper/config/schema/oe_theme_helper.schema.yml]: Since you are extending from the scale image effect, you can just call it type `image.effect.image_scale` and inherit its properties. And add your own `multiplier` property to the definition.

File: modules/oe_theme_helper/templates/retina-image-scale-summary.html.twig [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+ *   - upscale: If images larger than their original size can scale.
+ * - effect: The effect information, including:
+ *   - id: The effect identifier.
+ *   - label: The effect name.
+ *   - description: The effect description.
+ *
+ * @ingroup themeable
+ */
+#}
+{% if data.width and data.height -%}
```
Comment (upchuk) [inline modules/oe_theme_helper/templates/retina-image-scale-summary.html.twig]: Why is this needed? You copy pasted this entire thing and made no changes whatsoever..

File: tests/Kernel/BlockTest.php [hunk 3]
```diff
@@ -30,6 +30,7 @@ class BlockTest extends EntityKernelTestBase {
     'image',
     'breakpoint',
     'responsive_image',
+    'oe_theme_helper',
```
Comment (upchuk) [inline tests/Kernel/BlockTest.php]: Why?

File: modules/oe_theme_helper/src/Plugin/ImageEffect/RetinaScaleImageEffect.php [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+namespace Drupal\oe_theme_helper\Plugin\ImageEffect;
+
+use Drupal\Component\Utility\Image;
+use Drupal\Core\Form\FormStateInterface;
+use Drupal\Core\Image\ImageInterface;
+use Drupal\image\Plugin\ImageEffect\ScaleImageEffect;
+
+/**
+ * Scales an image resource.
```
Comment (upchuk) [inline modules/oe_theme_helper/src/Plugin/ImageEffect/RetinaScaleImageEffect.php]: Can you please please stop copy pasting comments and things if you don't update them to match what you are doing?

File: modules/oe_theme_helper/src/Plugin/ImageEffect/RetinaScaleImageEffect.php [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+class RetinaScaleImageEffect extends ScaleImageEffect {
+
+  /**
+   * {@inheritdoc}
+   */
+  public function applyEffect(ImageInterface $image) {
+    $target_with = $this->configuration['width'];
+    $target_height = $this->configuration['height'];
+    $upscale = $this->configuration['upscale'];
+    if (!$upscale) {
```
Comment (upchuk) [inline modules/oe_theme_helper/src/Plugin/ImageEffect/RetinaScaleImageEffect.php]: You can inline these. It's very difficult to read. Had to move them to one line to understand what is going on...

File: modules/oe_theme_helper/src/Plugin/ImageEffect/RetinaScaleImageEffect.php [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+      if (
+      (!empty($target_with) && $this->configuration['width'] > $image->getWidth()) ||
```
Comment (upchuk) [inline modules/oe_theme_helper/src/Plugin/ImageEffect/RetinaScaleImageEffect.php]: If you put `$this->configuration['width']` into a variable, why you checking it here like this again?

File: modules/oe_theme_helper/src/Plugin/ImageEffect/RetinaScaleImageEffect.php [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+      (!empty($target_height) && $this->configuration['height'] > $image->getHeight())
+      ) {
+        $target_with = $image->getWidth() * $this->configuration['multiplier'];
+        $target_height = $image->getHeight() * $this->configuration['multiplier'];
+        $upscale = TRUE;
```
Comment (upchuk) [inline modules/oe_theme_helper/src/Plugin/ImageEffect/RetinaScaleImageEffect.php]: I don't think we should override `$upscale` unless there is also a multiplier higher than 1 right?

File: modules/oe_theme_helper/src/Plugin/ImageEffect/RetinaScaleImageEffect.php [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+      ]);
+      return FALSE;
+    }
+    return TRUE;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function transformDimensions(array &$dimensions, $uri) {
```
Comment (upchuk) [inline modules/oe_theme_helper/src/Plugin/ImageEffect/RetinaScaleImageEffect.php]: See comments from `applyEffect ` for here as well

File: modules/oe_theme_helper/src/Plugin/ImageEffect/RetinaScaleImageEffect.php [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+      Image::scaleDimensions($dimensions, $target_with, $target_height, $upscale);
+    }
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getSummary() {
+    $summary = [
+      '#theme' => 'retina_image_scale_summary',
```
Comment (upchuk) [inline modules/oe_theme_helper/src/Plugin/ImageEffect/RetinaScaleImageEffect.php]: See other comment.

File: modules/oe_theme_helper/src/Plugin/ImageEffect/RetinaScaleImageEffect.php [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+
+  /**
+   * {@inheritdoc}
+   */
+  public function buildConfigurationForm(array $form, FormStateInterface $form_state) {
+    $form = parent::buildConfigurationForm($form, $form_state);
+    $form['multiplier'] = [
+      '#type' => 'number',
+      '#title' => t('Multiplier'),
+      '#default_value' => $this->configuration['multiplier'],
```
Comment (upchuk) [inline modules/oe_theme_helper/src/Plugin/ImageEffect/RetinaScaleImageEffect.php]: Let's put a description to say what this does and that it overrides the upscale etc.

File: modules/oe_theme_helper/config/schema/oe_theme_helper.schema.yml [hunk 11]
```diff
@@ -0,0 +1,10 @@
...
+  label: 'Retina Image scale'
```
Comment (upchuk) [inline modules/oe_theme_helper/config/schema/oe_theme_helper.schema.yml]: Also let's not call it "Retina" cause it has nothing to do with retina per se. It's a multiplier based. If I configure a multiplier of 10, it has nothing to do with retina. So I'd call the plugin `MultiplierImageScale`. Though it's also only upscales if the image is smaller than the configuration. Any ideas for naming?

File: modules/oe_theme_helper/config/schema/oe_theme_helper.schema.yml (hunk 1 above)
Comment (imanoleguskiza) [reply modules/oe_theme_helper/config/schema/oe_theme_helper.schema.yml]: I am extending the scale image effect but I'm also creating a new plugin with a separate id. Correct me if I'm wrong but if I use the "image.effect.image_scale" type then the multiplier property would apply to configuration using the "image_scale" id.

File: modules/oe_theme_helper/templates/retina-image-scale-summary.html.twig (hunk 2 above)
Comment (imanoleguskiza) [reply modules/oe_theme_helper/templates/retina-image-scale-summary.html.twig]: The twig template is needed in because the image style preview requires it. Since the changes we made to the base scale image style are a very edge case I didn't think we would need to change anything on the summary template either.

Comment (upchuk) [reply modules/oe_theme_helper/templates/retina-image-scale-summary.html.twig]: Then why not use the original?!

File: modules/oe_theme_helper/config/schema/oe_theme_helper.schema.yml (hunk 1 above)
Comment (upchuk) [reply modules/oe_theme_helper/config/schema/oe_theme_helper.schema.yml]: config schema "types" extend one another. So `image.effect.image_scale` extends `image_size` and adds its own config fields. You can extend from `image.effect.image_scale` and add yours (the multipler0

File: tests/Kernel/BlockTest.php (hunk 3 above)
Comment (imanoleguskiza) [reply tests/Kernel/BlockTest.php]: When installing the theme we install all the image styles. Most of the schema for this image styles comes from the image module (already there) but the new retine_image_scale is defined on the oe_theme_helper, hence the need to enable it as well.

File: modules/oe_theme_helper/config/schema/oe_theme_helper.schema.yml (hunk 11 above)
Comment (imanoleguskiza) [reply modules/oe_theme_helper/config/schema/oe_theme_helper.schema.yml]: We could change the configuration form so that the users are forced to select certain multipliers. Currently only 2x and 3x retina screens are available (after a quick google search, let me know if there are more options) so we could restrict the field to those two options. Would this make sense?

Comment (upchuk) [reply modules/oe_theme_helper/config/schema/oe_theme_helper.schema.yml]: Yes, true, makes sense what you say.

File: modules/oe_theme_helper/templates/retina-image-scale-summary.html.twig (hunk 2 above)
Comment (imanoleguskiza) [reply modules/oe_theme_helper/templates/retina-image-scale-summary.html.twig]: Fair enough, I'll re-use the original template.

File: modules/oe_theme_helper/src/Plugin/ImageEffect/RetinaScaleImageEffect.php (hunk 4 above)
Comment (imanoleguskiza) [reply modules/oe_theme_helper/src/Plugin/ImageEffect/RetinaScaleImageEffect.php]: I updated the comments on the plugin to more clearly reflect what we are doing.

File: modules/oe_theme_helper/src/Plugin/ImageEffect/RetinaScaleImageEffect.php (hunk 7 above)
Comment (imanoleguskiza) [reply modules/oe_theme_helper/src/Plugin/ImageEffect/RetinaScaleImageEffect.php]: Now that we are restricting the multiplier values we can keep this as it is.

File: modules/oe_theme_helper/src/Plugin/ImageEffect/RetinaScaleImageEffect.php (hunk 10 above)
Comment (imanoleguskiza) [reply modules/oe_theme_helper/src/Plugin/ImageEffect/RetinaScaleImageEffect.php]: Added a description.

File: modules/oe_theme_helper/config/schema/oe_theme_helper.schema.yml [hunk 12]
```diff
@@ -0,0 +1,7 @@
+image.effect.retina_image_scale:
+  type: image.effect.image_scale
+  label: 'Retina Image scale'
+  mapping:
+    multiplier:
+      type: integer
+      label: 'Multiplier'
```
Comment (dxvargas) [inline modules/oe_theme_helper/config/schema/oe_theme_helper.schema.yml]: Needs a new line

File: modules/oe_theme_helper/src/Plugin/ImageEffect/RetinaScaleImageEffect.php [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+    if ($dimensions['width'] && $dimensions['height']) {
+      $target_with = $this->configuration['width'];
+      $target_height = $this->configuration['height'];
+      // If we are not upscaling the image, check to see if it's smaller
+      // than the defined dimensions.
+      if (!$upscale = $this->configuration['upscale']) {
+        if (
+          (!empty($target_with) && $this->configuration['width'] > $dimensions['width']) ||
+          (!empty($target_height) && $this->configuration['height'] > $dimensions['height'])
+        ) {
```
Comment (dxvargas) [inline modules/oe_theme_helper/src/Plugin/ImageEffect/RetinaScaleImageEffect.php]: In these conditions it would be better to use:
`$target_with` instead of `$this->configuration['width']`
`$target_height` instead of `$this->configuration['height']`

File: modules/oe_theme_helper/src/Plugin/ImageEffect/RetinaScaleImageEffect.php (hunk 5 above)
Comment (upchuk) [reply modules/oe_theme_helper/src/Plugin/ImageEffect/RetinaScaleImageEffect.php]: @imanoleguskiza You didn't do this.

File: modules/oe_theme_helper/src/Plugin/ImageEffect/RetinaScaleImageEffect.php [hunk 14]
```diff
@@ -0,0 +1,132 @@
...
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function buildConfigurationForm(array $form, FormStateInterface $form_state) {
+    $form = parent::buildConfigurationForm($form, $form_state);
+    $form['multiplier'] = [
+      '#type' => 'select',
+      '#title' => t('Multiplier'),
```
Comment (upchuk) [inline modules/oe_theme_helper/src/Plugin/ImageEffect/RetinaScaleImageEffect.php]: `$this->t()` everywhere, not `t()`.

File: modules/oe_theme_helper/tests/src/Functional/RetinaScaleEffectTest.php [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+
+use Drupal\FunctionalTests\Image\ToolkitTestBase;
+
+/**
+ * Tests that the Retina Scale effect upscales images appropriately.
+ *
+ * @group image
+ */
+class RetinaScaleEffectTest extends ToolkitTestBase {
+  /**
```
Comment (upchuk) [inline modules/oe_theme_helper/tests/src/Functional/RetinaScaleEffectTest.php]: Empty line and `inheritdoc`

File: modules/oe_theme_helper/tests/src/Functional/RetinaScaleEffectTest.php [hunk 16]
```diff
@@ -0,0 +1,88 @@
+<?php
+
+declare(strict_types = 1);
```
Comment (upchuk) [inline modules/oe_theme_helper/tests/src/Functional/RetinaScaleEffectTest.php]: Missing return types on all methods.

File: modules/oe_theme_helper/tests/src/Functional/RetinaScaleEffectTest.php [hunk 17]
```diff
@@ -0,0 +1,88 @@
...
+
+  /**
+   * Asserts the effect processing of an image effect plugin.
+   *
+   * @param string $effect_name
+   *   The name of the image effect to test.
+   * @param array $data
+   *   The data to pass to the image effect.
+   *
+   * @return bool
```
Comment (upchuk) [inline modules/oe_theme_helper/tests/src/Functional/RetinaScaleEffectTest.php]: Not a bool

File: modules/oe_theme_helper/tests/src/Functional/RetinaScaleEffectTest.php [hunk 18]
```diff
@@ -0,0 +1,88 @@
...
+namespace Drupal\Tests\oe_theme_helper\Functional;
+
+use Drupal\FunctionalTests\Image\ToolkitTestBase;
+
+/**
+ * Tests that the Retina Scale effect upscales images appropriately.
+ *
+ * @group image
+ */
+class RetinaScaleEffectTest extends ToolkitTestBase {
```
Comment (upchuk) [inline modules/oe_theme_helper/tests/src/Functional/RetinaScaleEffectTest.php]: You are only testing the multiplier. What if the image is is bigger than the config. Or if the upscale is checked?

File: modules/oe_theme_helper/src/Plugin/ImageEffect/RetinaScaleImageEffect.php (hunk 14 above)
Comment (imanoleguskiza) [reply modules/oe_theme_helper/src/Plugin/ImageEffect/RetinaScaleImageEffect.php]: Done.

File: modules/oe_theme_helper/config/schema/oe_theme_helper.schema.yml (hunk 12 above)
Comment (imanoleguskiza) [reply modules/oe_theme_helper/config/schema/oe_theme_helper.schema.yml]: Done.

File: modules/oe_theme_helper/src/Plugin/ImageEffect/RetinaScaleImageEffect.php (hunk 5 above)
Comment (imanoleguskiza) [reply modules/oe_theme_helper/src/Plugin/ImageEffect/RetinaScaleImageEffect.php]: Done.

File: modules/oe_theme_helper/tests/src/Functional/RetinaScaleEffectTest.php (hunk 15 above)
Comment (imanoleguskiza) [reply modules/oe_theme_helper/tests/src/Functional/RetinaScaleEffectTest.php]: Done

File: modules/oe_theme_helper/tests/src/Functional/RetinaScaleEffectTest.php (hunk 16 above)
Comment (imanoleguskiza) [reply modules/oe_theme_helper/tests/src/Functional/RetinaScaleEffectTest.php]: Done

File: modules/oe_theme_helper/tests/src/Functional/RetinaScaleEffectTest.php (hunk 18 above)
Comment (imanoleguskiza) [reply modules/oe_theme_helper/tests/src/Functional/RetinaScaleEffectTest.php]: Tests added

File: modules/oe_theme_helper/tests/src/Functional/RetinaScaleEffectTest.php [hunk 19]
```diff
@@ -0,0 +1,115 @@
...
+  /**
+   * {@inheritdoc}
+   */
+  protected function setUp(): void {
+    parent::setUp();
+    $this->manager = $this->container->get('plugin.manager.image.effect');
+  }
+
+  /**
+   * Test the retina_image_scale_effect() function.
```
Comment (upchuk) [inline modules/oe_theme_helper/tests/src/Functional/RetinaScaleEffectTest.php]: What function is this?

File: modules/oe_theme_helper/tests/src/Functional/RetinaScaleEffectTest.php [hunk 20]
```diff
@@ -0,0 +1,115 @@
...
+   */
+  public function testRetinaScaleEffect(): void {
+    $this->assertImageEffect('retina_image_scale', [
+      // Set the desired width to be much higher than the image width.
```
Comment (upchuk) [inline modules/oe_theme_helper/tests/src/Functional/RetinaScaleEffectTest.php]: `. By default, it uses a multiplier of 2.`

File: modules/oe_theme_helper/tests/src/Functional/RetinaScaleEffectTest.php [hunk 21]
```diff
@@ -0,0 +1,115 @@
...
+      'width' => $this->image->getWidth() * 10,
+    ]);
+    $this->assertToolkitOperationsCalled(['scale']);
+
+    // Check the parameters.
```
Comment (upchuk) [inline modules/oe_theme_helper/tests/src/Functional/RetinaScaleEffectTest.php]: Can remove this comment.

File: modules/oe_theme_helper/tests/src/Functional/RetinaScaleEffectTest.php [hunk 22]
```diff
@@ -0,0 +1,115 @@
...
+    $calls = $this->imageTestGetAllCalls();
+    $this->assertEqual($calls['scale'][0][0], $this->image->getWidth() * 2, 'Width is double the original size.');
+  }
+
+  /**
+   * Test the image_scale_effect() function using a multiplier of 3.
```
Comment (upchuk) [inline modules/oe_theme_helper/tests/src/Functional/RetinaScaleEffectTest.php]: What function?

File: modules/oe_theme_helper/tests/src/Functional/RetinaScaleEffectTest.php [hunk 23]
```diff
@@ -0,0 +1,115 @@
...
+   */
+  public function testTripleMultiplierRetinaScaleEffect(): void {
+    $this->assertImageEffect('retina_image_scale', [
+      // Set the desired width to be much higher than the image width.
+      'width' => $this->image->getWidth() * 10,
+      // Set a multiplier of three for the test.
```
Comment (upchuk) [inline modules/oe_theme_helper/tests/src/Functional/RetinaScaleEffectTest.php]: No need for this comment.

File: modules/oe_theme_helper/tests/src/Functional/RetinaScaleEffectTest.php [hunk 24]
```diff
@@ -0,0 +1,115 @@
...
+      'multiplier' => 3,
+    ]);
+    $this->assertToolkitOperationsCalled(['scale']);
+
+    // Check the parameters.
```
Comment (upchuk) [inline modules/oe_theme_helper/tests/src/Functional/RetinaScaleEffectTest.php]: Remove comment

File: modules/oe_theme_helper/tests/src/Functional/RetinaScaleEffectTest.php [hunk 25]
```diff
@@ -0,0 +1,115 @@
...
+    $calls = $this->imageTestGetAllCalls();
+    $this->assertEqual($calls['scale'][0][0], $this->image->getWidth() * 3, 'Width is triple the original size.');
+  }
+
+  /**
+   * Test the image_scale_effect() function using a big enough image.
```
Comment (upchuk) [inline modules/oe_theme_helper/tests/src/Functional/RetinaScaleEffectTest.php]: What function?

File: modules/oe_theme_helper/tests/src/Functional/RetinaScaleEffectTest.php [hunk 26]
```diff
@@ -0,0 +1,115 @@
...
+   */
+  public function testScaleEffect(): void {
+    $this->assertImageEffect('image_scale', [
+      // Set the image width to be smaller than the image width.
```
Comment (upchuk) [inline modules/oe_theme_helper/tests/src/Functional/RetinaScaleEffectTest.php]: Comment sounds weird. Maybe keep it in line with the same comment in the previous methods?

File: modules/oe_theme_helper/tests/src/Functional/RetinaScaleEffectTest.php [hunk 27]
```diff
@@ -0,0 +1,115 @@
...
+    // Check the parameters.
+    $calls = $this->imageTestGetAllCalls();
+    $this->assertEqual($calls['scale'][0][0], $this->image->getWidth() * 3, 'Width is triple the original size.');
+  }
+
+  /**
+   * Test the image_scale_effect() function using a big enough image.
+   */
+  public function testScaleEffect(): void {
+    $this->assertImageEffect('image_scale', [
```
Comment (upchuk) [inline modules/oe_theme_helper/tests/src/Functional/RetinaScaleEffectTest.php]: Are we testing the `image_scale` effect or the `retina_image_scale` effect?

File: modules/oe_theme_helper/tests/src/Functional/RetinaScaleEffectTest.php [hunk 28]
```diff
@@ -0,0 +1,115 @@
...
+    $this->assertToolkitOperationsCalled(['scale']);
+
+    // Check the parameters.
+    $calls = $this->imageTestGetAllCalls();
+    $this->assertEqual($calls['scale'][0][0], 10, 'Width was passed correctly');
+    $this->assertEqual($calls['scale'][0][1], 10, 'Height was based off aspect ratio and passed correctly');
+  }
+
+  /**
+   * Test the image_scale_effect() function using upscaling.
```
Comment (upchuk) [inline modules/oe_theme_helper/tests/src/Functional/RetinaScaleEffectTest.php]: Again, copy paste from core I see...

File: modules/oe_theme_helper/tests/src/Functional/RetinaScaleEffectTest.php [hunk 29]
```diff
@@ -0,0 +1,115 @@
...
+   */
+  public function testScaleEffectWithUpscaling(): void {
+    $this->assertImageEffect('image_scale', [
```
Comment (upchuk) [inline modules/oe_theme_helper/tests/src/Functional/RetinaScaleEffectTest.php]: What are we testing here?

File: modules/oe_theme_helper/src/Plugin/ImageEffect/RetinaScaleImageEffect.php [hunk 30]
```diff
@@ -0,0 +1,128 @@
...
+    $form['multiplier'] = [
+      '#type' => 'select',
+      '#title' => $this->t('Multiplier'),
+      '#options' => [
+        2 => '2x',
+        3 => '3x',
+      ],
+      '#default_value' => $this->configuration['multiplier'],
+      '#required' => TRUE,
+      '#description' => $this->t("The image will be upscaled according to this multiplier if it is smaller than the dimensions defined in the 'Width' and 'Height' properties."),
```
Comment (upchuk) [inline modules/oe_theme_helper/src/Plugin/ImageEffect/RetinaScaleImageEffect.php]: Single quotes.

Also we should mention that it only gets *upscaled* if the `upscale` is not unchecked no?

File: modules/oe_theme_helper/tests/src/Functional/RetinaScaleEffectTest.php [hunk 31]
```diff
@@ -0,0 +1,115 @@
...
+   * The image effect manager.
+   *
+   * @var \Drupal\image\ImageEffectManager
+   */
+  protected $manager;
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function setUp(): void {
```
Comment (upchuk) [inline modules/oe_theme_helper/tests/src/Functional/RetinaScaleEffectTest.php]: `setUp` is not ours so we shouldn't really put the void.
